NetworkSetFactsModule for HPE OneView

fgbulsoni opened the pull request:
##### SUMMARY
Added new oneview_network_set_facts module for retrieving [HPE OneView Network Set](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html#rest/network-sets) resource and unit tests.

Also updated oneview_module_loader to use short GPL3 copyright header.

Related issue for more information: [HPE OneView support](https://github.com/ansible/ansible/issues/28354)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`oneview_network_set_facts`

##### ANSIBLE VERSION
```
ansible 2.4.0 (hpe-oneview/network-set-facts b40e4077f6) last updated 2017/08/28 17:33:00 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/dev/core_ansible_related/ansible/lib/ansible
  executable location = /home/vagrant/dev/core_ansible_related/ansible/bin/ansible
  python version = 2.7.9 (default, Aug 21 2017, 11:24:49) [GCC 4.8.4]

```


##### ADDITIONAL INFORMATION
Example of usage:
```
- name: Gather facts about all Network Sets
  oneview_network_set_facts:
    config: /etc/oneview/oneview_config.json
  delegate_to: localhost

- debug: var=network_sets

- name: Gather paginated, filtered, and sorted facts about Network Sets
  oneview_network_set_facts:
    config: /etc/oneview/oneview_config.json
    params:
      start: 0
      count: 3
      sort: 'name:descending'
      filter: name='netset001'
  delegate_to: localhost

- debug: var=network_sets

- name: Gather facts about all Network Sets, excluding Ethernet networks
  oneview_network_set_facts:
    config: /etc/oneview/oneview_config.json
    options:
        - withoutEthernet
  delegate_to: localhost

- debug: var=network_sets


- name: Gather facts about a Network Set by name
  oneview_network_set_facts:
    config: /etc/oneview/oneview_config.json
    name: Name of the Network Set
  delegate_to: localhost

- debug: var=network_sets


- name: Gather facts about a Network Set by name, excluding Ethernet networks
  oneview_network_set_facts:
    config: /etc/oneview/oneview_config.json
    name: Name of the Network Set
    options:
        - withoutEthernet
  delegate_to: localhost

- debug: var=network_sets
```

fgbulsoni committed:
Add NetworkSetFactsModule for retrieving HPE OneView Network Sets

- Allow querying for Network Set resources in HPE OneView
- Adds unit tests to new module
- Updates oneview_module_loader copyright header to short GPL3 version

dagwieers commented:
I am not convinced there is a need for all these facts modules.

Wouldn't it be easier to have a single network facts module ?

dagwieers commented:
Also wrt. the *config.json* file, what is inside this file ?

Because if it is used for what I think it is being used, I am not convinced that this is the best way to support your modules. If this file is there to simply make it easier to provide hostname/username/password etc. yo your individual modules, this is the wrong way to do that. Using YAML anchors and references is much more convenient, and works across Ansible for every module.

dagwieers commented:
Talk to me on IRC #ansible-devel or @dag to discuss ASAP.

fgbulsoni committed:
Adding possibility to pass in credentials as parameters

fgbulsoni commented:
> Also wrt. the config.json file, what is inside this file ?

[Example of config json](https://github.com/HewlettPackard/oneview-ansible/blob/master/examples/oneview_config-rename.json)



> I am not convinced there is a need for all these facts modules.
> 
> Wouldn't it be easier to have a single network facts module ?

While it might reduce a few fact modules to centralize the networks, for instance, it'd make it so that the user would not know what to expect when querying for a resource. We've tried to keep the modules as tightly coupled to the [REST API documentation](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html) as possible, so that the users can easily check out that documentation and know what to expect for each version of OneView.

So, at the moment, most resources have a `ResourceNameModule` to manage it, handling states that modify it, such as `present`, `absent`, etc, and there's a `ResourceNameFactsModule` to retrieve information on the resource, basically having all of its `GET`s.

The data returned does get changed from API version to API version and ideally should be consulted on those docs.

fgbulsoni commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -202,7 +202,12 @@ class OneViewModuleBase(object):
     ONEVIEW_COMMON_ARGS = dict(
-        config=dict(required=False, type='str')
+        config=dict(required=False, type='path'),
+        hostname=dict(required=False, type='str'),
```
@dagwieers Is this the expected/wanted solution to not use a `json` file?
The task ended up looking like:
```yaml
    - name: Gather facts about all Network Sets
      oneview_network_set_facts:
        hostname: 172.16.101.48
        username: administrator
        password: my_password
        api_version: 500
      no_log: true
      delegate_to: localhost
```

I also had to use no_log otherwise the password was being logged and outputted to the screen when using `-vvv` during playbook execution.

dagwieers replied:
Correct ! no_log is needed for password parameter

dagwieers replied:
Again `required=False` is not needed here. Only add it when it is required.

fgbulsoni committed:
Removed required false and changed format of filter_by_name declaration

dagwieers commented:
@fgbulsoni It's your call to what interface the users will be using. If you think more facts modules is what it ought to be, you'll have to support it ;-) But if the interface changes with API versions, it means that people will have to rewrite playbooks when they move from API version.

Make that very clear in all the module docs, it seems essential that users know this upfront when they decide what version to work against.

dagwieers reviewed (approved):
Looking good mostly.

fgbulsoni commented:
Hey @dagwieers , I've updated the examples, removed the `required=False` and added documentation on the API version with the last commit.
Can you check if all is in there now? I *think* the PR is ready.

fgbulsoni committed:
Updated examples in docs to reflect new way to pass in credentials

- All examples of the oneview_network_set_facts updated to use
credential parameters
- All required=False from oneview base module removed
- Shared docs updated to bring attention to API version being used

dagwieers commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -257,7 +257,13 @@ def _check_hpe_oneview_sdk(self):
 
     def _create_oneview_client(self):
-        if not self.module.params['config']:
+        if self.module.params.get('hostname'):
```
So I am not sure this is the best way to do this. I would simply look for each of the variables at the various locaions, and merge them all together, using a precedence rule: parameter over config over environment.

But this is something you can add later. Not something you have to worry about now.
